// @nativescript/angular/esm2022/lib/cdk/dialog/dialog-module.mjs
import { NgModule } from '@angular/core';
import { NativeDialogCloseDirective } from './dialog-content-directives';
import { NativeDialogService } from './dialog-services';
import * as i0 from "@angular/core";
export class NativeDialogModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.1.0", ngImport: i0, type: NativeDialogModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "18.1.0", ngImport: i0, type: NativeDialogModule, imports: [NativeDialogCloseDirective], exports: [NativeDialogCloseDirective] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "18.1.0", ngImport: i0, type: NativeDialogModule, providers: [NativeDialogService] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.1.0", ngImport: i0, type: NativeDialogModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [NativeDialogCloseDirective],
                    exports: [NativeDialogCloseDirective],
                    providers: [NativeDialogService],
                }]
        }] });
