// @nativescript/angular/esm2022/lib/legacy/router/ns-empty-outlet.component.mjs
import { Component, ViewChild } from '@angular/core';
import { Page } from '@nativescript/core';
import { PageRouterOutlet } from './page-router-outlet';
import * as i0 from "@angular/core";
import * as i1 from "@nativescript/core";
export class NSEmptyOutletComponent {
    constructor(page) {
        this.page = page;
        if (this.page) {
            this.page.actionBarHidden = true;
            this.page.on('loaded', () => {
                if (this.pageRouterOutlet && this.page.frame) {
                    this.pageRouterOutlet.setActionBarVisibility(this.page.frame.actionBarVisibility);
                }
            });
        }
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.1.0", ngImport: i0, type: NSEmptyOutletComponent, deps: [{ token: i1.Page }], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "18.1.0", type: NSEmptyOutletComponent, isStandalone: true, selector: "ns-empty-outlet", viewQueries: [{ propertyName: "pageRouterOutlet", first: true, predicate: PageRouterOutlet, descendants: true, read: PageRouterOutlet }], ngImport: i0, template: "<page-router-outlet isEmptyOutlet='true'></page-router-outlet>", isInline: true, dependencies: [{ kind: "directive", type: PageRouterOutlet, selector: "page-router-outlet", outputs: ["activate", "deactivate"] }] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.1.0", ngImport: i0, type: NSEmptyOutletComponent, decorators: [{
            type: Component,
            args: [{
                    // tslint:disable-next-line:component-selector
                    selector: 'ns-empty-outlet',
                    template: "<page-router-outlet isEmptyOutlet='true'></page-router-outlet>",
                    standalone: true,
                    imports: [PageRouterOutlet],
                }]
        }], ctorParameters: () => [{ type: i1.Page }], propDecorators: { pageRouterOutlet: [{
                type: ViewChild,
                args: [PageRouterOutlet, { read: PageRouterOutlet, static: false }]
            }] } });
