// @nativescript/angular/esm2022/lib/cdk/platform-filters/ios-filter.component.mjs
/* eslint-disable @angular-eslint/component-selector */
import { Component, Inject } from '@angular/core';
import { platformNames } from '@nativescript/core';
import { DEVICE } from '../../tokens';
import * as i0 from "@angular/core";
export class IOSFilterComponent {
    constructor(device) {
        this.show = device.os === platformNames.ios;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.1.0", ngImport: i0, type: IOSFilterComponent, deps: [{ token: DEVICE }], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.0.0", version: "18.1.0", type: IOSFilterComponent, isStandalone: true, selector: "ios", ngImport: i0, template: `@if (show) { 
    <ng-content></ng-content>
  }`, isInline: true }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.1.0", ngImport: i0, type: IOSFilterComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'ios',
                    template: `@if (show) { 
    <ng-content></ng-content>
  }`,
                    standalone: true,
                }]
        }], ctorParameters: () => [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [DEVICE]
                }] }] });
