// @nativescript/angular/esm2022/lib/cdk/platform-filters/vision-filter.component.mjs
/* eslint-disable @angular-eslint/component-selector */
import { Component, Inject } from '@angular/core';
import { platformNames } from '@nativescript/core';
import { DEVICE } from '../../tokens';
import * as i0 from "@angular/core";
export class VisionOSFilterComponent {
    constructor(device) {
        // Note: casting any to be backwards compatible
        // Available in 8.6+ only
        this.show = device.os === platformNames.visionos;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.1.0", ngImport: i0, type: VisionOSFilterComponent, deps: [{ token: DEVICE }], target: i0.ɵɵFactoryTarget.Component }); }
    static { this.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.0.0", version: "18.1.0", type: VisionOSFilterComponent, isStandalone: true, selector: "visionos", ngImport: i0, template: `@if (show) { 
    <ng-content></ng-content>
  }`, isInline: true }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.1.0", ngImport: i0, type: VisionOSFilterComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'visionos',
                    template: `@if (show) { 
    <ng-content></ng-content>
  }`,
                    standalone: true,
                }]
        }], ctorParameters: () => [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [DEVICE]
                }] }] });
